design/0003-byop.md

```diff
@@ -0,0 +1,127 @@
+- Linkerd Issue: [linkerd/linkerd2#3590](https://github.com/linkerd/linkerd2/issues/3590)
+- Reviewers: @grampelberg @alpeb @siggy
+
+# Summary
+
+[summary]: #summary
+
+This RFC aims to move prometheus as an add-on i.e not mandated
+but adviced. This includes changes to the linkerd-prometheus manifests to make it as a add-on and allowing more configuration (more on this later),
+along with documentation and tooling to use existing prometheus installation.
+
+# Problem Statement
+
+[problem-statement]: #problem-statement
+
+Lots of users already have a exisiting prometheus installation in their cluster and don't like managing
+another prometheus instance especially because they could be resource intensive and painful. They usually
+have a prometheus installation paired up with a long term monitroing solutions like Cortex,InfluxDB, or third party
+solutions like Grafana Cloud, etc.
+In those cases prometheus is just a temporary component, that keeps pushing metrics into their long term solution, making users want to re-use their exisitng prometheus.
+
+The following are the list of problems:
+
+- Hard requirement of linkerd-prometheus.
+
+- Configuring the current linkerd-prometheus for remote-writes, etc is hard, because the cm is overriden after upgrades, etc.
+
+- Not enough documentation for bringing your own prometheus case.
+
+Once the RFC is implemented, Apart from the default prometheus enabled installation Users should be
+able to use their exising prometheus instance, to also scrape metrics from the proxy
+
+# Design proposal
+
+[design-proposal]: #design-proposal
+
+## Prometheus as an Add-On
+
+By moving Prometheus as an add-on, We can make it as an optional component that can be used both through Helm and Linkerd CLI.
+The add-on model would well with upgrades, etc as per [#3955](https://github.com/linkerd/linkerd2/pull/3955). This is done by
+moving the prometheus components into a separate sub-chart in the linkerd repo
+and updating the code. The same health checks are also expected, but as a separate
+check Category, which would be transient i.e runs only when it recognises
+that the add-on is enabled.
+
+By default, the prometheus add-on will be **enabled**.
+
+This also provides a single place of configuration for all things related to
+linkerd-prometheus. Though we are making linkerd-prometheus optional but having it is great for Linkerd and external users, for reasons like separation of
+concerns, etc. By allowing the `remote-write`, `alerts`, etc configuration in the
+linkerd-prometheus, we would decrease the need/want for external users to tie their existing prometheus to Linkerd,
+as it would be even easier for them to configure `remote-write`, etc
+configuration from their exisiting prometheus to the linkerd-prometheus right from the start.
+
+The following default prometheus configuration is planned:
+
+```yaml
+prometheus:
+  enabled: true
+  name: linkerd-prometheus
+  url: 
+  image: 
+  global:
```

Comment:
Isn't it possible to load all these from files? Then folks could just manage configmaps and not actually change the chart in any way.

Reply:
I think the first step is to extract our current Prometheus configmap into a separate yaml template (currently it lives together with the prom service and deployment). That configmap is currently being reworked in linkerd/linkerd2#4220, which adds the ability to plug-in recording and alerting rules through separate configmaps (no other prom configs items are pluggable like that from what I can tell).
Then this new Prometheus addon `values.yaml` could just have a few minimal entries like `enabled`, `image`, etc, and the new entries from that PR to configure rules. We could also have an entry to entirely override the provided Prometheus configmap with an external one, in which case the user would need to be advised to follow the scrape configs from the default configmap.
Makes sense?

Reply:
As @alpeb mentioned, I integrated stuff from https://github.com/linkerd/linkerd2/pull/4220 into this `prometheus` field. It dosen't help a lot, as only rules, etc can be added as files and other configuration like `remote_write`, `alerting` etc feels better to be taken directly.

> Then this new Prometheus addon values.yaml could just have a few minimal entries like enabled, image, etc, and the new entries from that PR to configure rules.

Users can always choose to ignore fields that he dont want to fill!

> We could also have an entry to entirely override the provided Prometheus configmap with an external one, in which case the user would need to be advised to follow the scrape configs from the default configmap.

This is interesting but it essentially feels like a BYOP case? and requires complication on maintaing the linkerd-addons configmap.